add a test that shows the preservation UDP traffic when server pod cycles

JacobTanenbaum opened the pull request:
When using NodePort services and long running connections that stale conntrack
entries on pod deletion can halt the flow of traffic. Add a test case to check
that conntrack entries are cleaned up.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind cleanup


**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
adds additional testing to ensure that udp pods conntrack are cleaned up 
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dcbw commented:
/ok-to-test

dcbw commented on test/e2e/network/service.go:
```diff
@@ -1130,6 +1234,50 @@ var _ = SIGDescribe("Services", func() {
+		err = <-errorChannel
+
+		framework.ExpectNoError(err, "pod communication failed")
+
```
@JacobTanenbaum unncessary newline

JacobTanenbaum replied:
fixed

dcbw commented on test/e2e/network/service.go:
```diff
@@ -1130,6 +1234,50 @@ var _ = SIGDescribe("Services", func() {
 
+	/*
+	   Release : v1.19
+	   Testname: Service, perserve UDP traffice when server backing a service cycles
```
"traffic"
server -> pod

JacobTanenbaum replied:
fixec

dcbw commented on test/e2e/network/service.go:
```diff
@@ -1130,6 +1234,50 @@ var _ = SIGDescribe("Services", func() {
+	   Description: Create a UDP ClusterIP Service with a backing server pod. Run traffic
+	   through the service to the server pod. The traffic flow must recover when the server
+	   pod originally serving requests is deleted and a new pod is ready to serve requests
+
```
Unecessary newline

JacobTanenbaum replied:
fixed

dcbw commented on test/e2e/network/service.go:
```diff
@@ -1478,6 +1626,7 @@ var _ = SIGDescribe("Services", func() {
 		e2eservice.TestReachableHTTP(nodeIP, tcpNodePort, e2eservice.KubeProxyLagTimeout)
 
 		ginkgo.By("hitting the UDP service's NodePort")
+		ginkgo.By("creating a UDP service " + serviceName + " with type=ClusterIP in namespace " + ns2)
```
Leftover debugging code?

JacobTanenbaum replied:
removed the code

dcbw commented:
@JacobTanenbaum also stuff changed upstream versus your code:

```
W0415 18:49:36.619] compilepkg: error running subcommand: exit status 2
I0415 18:49:36.720] /bazel-scratch/.cache/bazel/_bazel_prow/48d5366022b4e3197674c8d6e2bee219/sandbox/processwrapper-sandbox/7799/execroot/io_k8s_kubernetes/test/e2e/network/service.go:1249:18: f.NewAgnhostPod undefined (type *"k8s.io/kubernetes/test/e2e/framework".Framework has no field or method NewAgnhostPod)
I0415 18:49:36.721] /bazel-scratch/.cache/bazel/_bazel_prow/48d5366022b4e3197674c8d6e2bee219/sandbox/processwrapper-sandbox/7799/execroot/io_k8s_kubernetes/test/e2e/network/service.go:1253:28: f.WaitForPodReady undefined (type *"k8s.io/kubernetes/test/e2e/framework".Framework has no field or method WaitForPodReady)
I0415 18:49:36.722] /bazel-scratch/.cache/bazel/_bazel_prow/48d5366022b4e3197674c8d6e2bee219/sandbox/processwrapper-sandbox/7799/execroot/io_k8s_kubernetes/test/e2e/network/service.go:1258:18: f.NewAgnhostPod undefined (type *"k8s.io/kubernetes/test/e2e/framework".Framework has no field or method NewAgnhostPod)
I0415 18:49:36.722] /bazel-scratch/.cache/bazel/_bazel_prow/48d5366022b4e3197674c8d6e2bee219/sandbox/processwrapper-sandbox/7799/execroot/io_k8s_kubernetes/test/e2e/network/service.go:1262:28: f.WaitForPodReady undefined (type *"k8s.io/kubernetes/test/e2e/framework".Framework has no field or method WaitForPodReady)
I0415 18:49:36.823] make: Leaving directory '/go/src/k8s.io/kubernetes'
W0415 18:49:36.924] Target //build/release-tars:release-tars failed to build
```

JacobTanenbaum committed:
Add a test that checks if conntrack entries are cleaned up for UDP traffic

ensure that when a pod servicing UDP traffic is deleted the conntrack entries
are cleaned up and another backend can pick up the traffic  with minimal
interruption

When using NodePort services and long running connections that on pod deletion
stale conntrack entries can halt the flow of traffic. Add a test case to check
that conntrack entries are cleaned up.

JacobTanenbaum commented:
/retest

dcbw commented:
@JacobTanenbaum 

```
+++ Running case: verify.spelling 
+++ working dir: /home/prow/go/src/k8s.io/kubernetes
+++ command: bash "hack/make-rules/../../hack/verify-spelling.sh"
test/e2e/network/service.go:556:19: "Continous" is a misspelling of "Continuous"
test/e2e/network/service.go:614:18: "Continous" is a misspelling of "Continuous"
test/e2e/network/service.go:1232:23: "perserve" is a misspelling of "preserve"
+++ exit code: 1
+++ error: 1
[0;31mFAILED[0m   verify-spelling.sh	8s
```

```
+++ Running case: verify.staticcheck 
+++ working dir: /home/prow/go/src/k8s.io/kubernetes
+++ command: bash "hack/make-rules/../../hack/verify-staticcheck.sh"
installing staticcheck from vendor
Errors from staticcheck:
test/e2e/network/service.go:536:29: Printf format %s reads arg #1, but call has only 0 args (SA5009)
```

```
+++ Running case: verify.govet 
+++ working dir: /home/prow/go/src/k8s.io/kubernetes
+++ command: bash "hack/make-rules/../../hack/verify-govet.sh"
NOTE: hack/make-rules/../../hack/verify-govet.sh has been replaced by 'make vet'

The equivalent of this invocation is: 
    make vet 


# k8s.io/kubernetes/test/e2e/network
test/e2e/network/service.go:536:19: Errorf format %s reads arg #1, but call has 0 args
make[1]: *** [Makefile:386: vet] Error 1
+++ exit code: 2
+++ error: 1
[0;31mFAILED[0m   verify-govet.sh	563s
```

aojea commented on test/e2e/network/service.go:
```diff
@@ -518,6 +518,110 @@ func pokeUDP(host string, port int, request string, params *UDPPokeParams) UDPPo
+
+	ret := UDPPokeResult{}
+
+	// Sanity check inputs, because it has happened.  These are the only things
```
nit , should we do the sanity check at the beginning, before initializing the variables?

```
hostPort := net.JoinHostPort(host, strconv.Itoa(port))
url := fmt.Sprintf("udp://%s", hostPort)
```

JacobTanenbaum replied:
added the sanity check above the variable initialization

aojea commented on test/e2e/network/service.go:
```diff
@@ -1123,6 +1227,48 @@ var _ = SIGDescribe("Services", func() {
+		framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(f.ClientSet, serverPod1.Name, f.Namespace.Name, framework.PodStartTimeout))
+
+		errorChannel := make(chan error)
+		go continuousEcho(udpService.Spec.ClusterIP, 80, 3*time.Second, 20, errorChannel)
```
This assumes that the e2e.test binary has access to the clusterIP, right?

JacobTanenbaum replied:
@aojea  Yes, is that a bad assumption?

aojea replied:
I don't know if it is documented, but it doesn't seem a good idea if the e2e framework has to require this, I think that is not common that the test runner has access to the pod and clusterIP network.
I think that the best approach is to spawn another pod with the client `podClient` and execute it from the pod.
This e2e network performance test does something similar https://github.com/kubernetes/kubernetes/blob/2c2e9cdd72f363e6e6282f1ec2832fb6c2f4858a/test/e2e/network/networking_perf.go#L57-L61

You should verify this, but the agnost host has a `iperf` inside and maybe running in UDP mode can be valid for your test

aojea commented on test/conformance/testdata/conformance.yaml:
```diff
@@ -1644,6 +1644,15 @@
     the NodePort.
   release: v1.16
   file: test/e2e/network/service.go
+- testname: Service, preserve UDP traffic when pod backing a service cycles
```
There is a process to promote tests to conformance 
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-testing-in-kubernetes

> > The process to add new conformance tests is intended to decouple the development of useful tests from their promotion to conformance:
> 
> Contributors write and submit e2e tests, to be approved by owning SIGs
> Tests are proven to meet the conformance test requirements by review and by accumulation of data on flakiness and reliability
> A follow up PR is submitted to promote the test to conformance

aojea commented:
you have an example here to check for conntrack entries in a node, maybe it can be useful 
https://github.com/kubernetes/kubernetes/blob/d6f44b0ea9d624ba420de4e91f737957c1d148da/test/e2e/network/kube_proxy.go#L83-L88

aojea commented on test/e2e/network/service.go:
```diff
@@ -1123,6 +1228,50 @@ var _ = SIGDescribe("Services", func() {
 		framework.ExpectNoError(verifyServeHostnameServiceUp(cs, ns, host, podNames2, svc2IP, servicePort))
 	})
 
+	ginkgo.It("should be able to preserve UDP traffic when server pod cycles [Slow]", func() {
```
I'd name it `should be able to preserve UDP traffic when server pod cycles for NodePort service` to identify clearly this test. If we add a new one for ClusterIP services it can create confusion.
Also, we can remove the [Slow] tag now so we can check the test running in the CI, and come back later

JacobTanenbaum replied:
done

aojea commented on test/e2e/network/service.go:
```diff
@@ -1123,6 +1228,50 @@ var _ = SIGDescribe("Services", func() {
+
+		nodeIP, err := e2enode.PickIP(cs) // for later
+		framework.ExpectNoError(err)
+
```
`ginkgo.By("creating a UDP service " + serviceName + " with type=NodePort in namespace " + ns)`

JacobTanenbaum replied:
done

aojea commented on test/e2e/network/service.go:
```diff
@@ -1123,6 +1228,50 @@ var _ = SIGDescribe("Services", func() {
+		framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(f.ClientSet, serverPod1.Name, f.Namespace.Name, framework.PodStartTimeout))
+
+		errorChannel := make(chan error)
+		go continuousEcho(nodeIP, int(udpService.Spec.Ports[0].NodePort), 3*time.Second, 20, errorChannel)
```
Ginkgo and goroutines don't play nice and ask you to include `defer GinkgoRecover()` inside of your goroutine http://tonyyang132.blogspot.com/2018/08/ginkgo-and-goroutine.html.
We have to test this, it may break the CI randomly (I learn it the hard way :) )

JacobTanenbaum replied:
added `defer GinkgoRecover()`

aojea commented on test/e2e/network/service.go:
```diff
@@ -1123,6 +1228,50 @@ var _ = SIGDescribe("Services", func() {
+		framework.ExpectNoError(err)
+		framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(f.ClientSet, serverPod1.Name, f.Namespace.Name, framework.PodStartTimeout))
+
+		errorChannel := make(chan error)
```
`ginkgo.By(fmt.Sprintf("checking NodePort service %s on node name %s with public IP %s", serviceName, nodeName, nodeIP))`

JacobTanenbaum replied:
done

aojea commented on test/e2e/network/service.go:
```diff
@@ -1123,6 +1228,50 @@ var _ = SIGDescribe("Services", func() {
+		errorChannel := make(chan error)
+		go continuousEcho(nodeIP, int(udpService.Spec.Ports[0].NodePort), 3*time.Second, 20, errorChannel)
+
+		serverPod2 := newAgnhostPod(serverPod2Name, "netexec", fmt.Sprintf("--udp-port=%d", 80))
```
`ginkgo.By("creating a second pod for the service " + serviceName)`

JacobTanenbaum replied:
done

aojea commented on test/e2e/network/service.go:
```diff
@@ -1123,6 +1228,50 @@ var _ = SIGDescribe("Services", func() {
+		framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(f.ClientSet, serverPod1.Name, f.Namespace.Name, framework.PodStartTimeout))
+
+		errorChannel := make(chan error)
+		go continuousEcho(nodeIP, int(udpService.Spec.Ports[0].NodePort), 3*time.Second, 20, errorChannel)
```
can we avoid duplicating the pokeUDP() into a new function continuousEcho() and use it directly in a for loop?

```go
go func() {
     defer GinkgoRecover()

  	for i := 0; i < maxAttempts; i++ {
             pokeUDP(host, port, "echo hello", &UDPPokeParams{
			Timeout:  3 * time.Second,
			Response: "hello",
		})
		if result.Status != UDPSuccess {
			numErrors++
		}

       }
  	err = nil
	if numErrors >= 6 {
		err = fmt.Errorf("Too many Errors in continuous echo")
	}
	errorChannel <- err
}
```

JacobTanenbaum replied:
no, this wont exercise the behaviour every call to pokeUDP will generate a difference source port and create a different conntrack entry. I needed to modify pokeUDP so that continuousEcho() is the same stream of packets with the same source port, the service needs to correctly change the flow of traffic to the new pod

aojea replied:
ahh good point, thank you, totally missed that

aojea commented:
I have doubts about possible flakiness in this tests with current implementation, in busy environments the pods can take longer than expected and we are relying on the maxErrors thresholds to declare the test failed or OK, I don't know if we are going to be able to find an stable value for it

Let's test it in the CI removing the [Slow] tag and observe.
Maybe we can serialize it?

1. Create NodePort service with one pod
2. Test that UDP echo works (pokeUDP)
3. Delete current pod and create a new one
4. Test that UDP work (pokeUDP)

JacobTanenbaum commented:
@aojea I don't think this can be serialized. the issues with the above are are 
1. every call to pokeUDP() generate a new source port and a new conntrack entry 
2. if you delete the first backend pod before creating the second there is no guarantee that the second pod gets an IP address different from the the first backend pod

it does pass if you check the kubernetes-e2e-kind: [pull-kubernetes-e2e-kind ](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/90180/pull-kubernetes-e2e-kind/1258442935758753793)

@dcbw do you have thoughts on if this can be serialized?

JacobTanenbaum commented on test/e2e/network/service.go:
```diff
@@ -1123,6 +1228,50 @@ var _ = SIGDescribe("Services", func() {
+		})
+		framework.ExpectNoError(err)
+
+		serverPod1 := newAgnhostPod(serverPod1Name, "netexec", fmt.Sprintf("--udp-port=%d", 80))
```
done

aojea commented:
congrats, this looks promising in kind, let's see in gce :crossed_fingers: 

```
[It] should be able to preserve UDP traffic when server pod cycles for a NodePort service
  test/e2e/network/service.go:1232
[1mSTEP[0m: creating a UDP service clusterip-test with type=NodePort in services-43
[1mSTEP[0m: creating a backend pod for the service clusterip-test
[1mSTEP[0m: checking NodePort service clusterip-test on node with public IP 172.18.0.4
May  8 15:02:05.656: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May  8 15:02:07.724: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May  8 15:02:09.661: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May  8 15:02:11.684: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May  8 15:02:13.667: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May  8 15:02:15.701: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May  8 15:02:17.677: INFO: The status of Pod server-1 is Running (Ready = true)
[1mSTEP[0m: creating a second pod for the service clusterip-test
May  8 15:02:17.706: INFO: The status of Pod server-2 is Pending, waiting for it to be Running (with Ready = true)
May  8 15:02:18.690: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:19.705: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:19.735: INFO: The status of Pod server-2 is Pending, waiting for it to be Running (with Ready = true)
May  8 15:02:20.706: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:21.707: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:21.719: INFO: The status of Pod server-2 is Pending, waiting for it to be Running (with Ready = true)
May  8 15:02:22.708: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:23.709: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:23.716: INFO: The status of Pod server-2 is Pending, waiting for it to be Running (with Ready = true)
May  8 15:02:24.710: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:25.713: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:25.713: INFO: The status of Pod server-2 is Running (Ready = true)
May  8 15:02:25.713: INFO: Cleaning up server-1 pod
May  8 15:02:26.716: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:27.716: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:28.732: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:29.733: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:30.738: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:31.739: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:32.741: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:33.744: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:34.745: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:35.746: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:36.750: INFO: Continuous echo("udp://172.18.0.4:32090"): success
May  8 15:02:37.755: INFO: Continuous echo("udp://172.18.0.4:32090"): success
[AfterEach] [sig-network] Services
  test/e2e/framework/framework.go:175
```

JacobTanenbaum commented:
/retest

JacobTanenbaum commented:
/retest

aojea commented on test/e2e/network/service.go:
```diff
@@ -1123,6 +1229,53 @@ var _ = SIGDescribe("Services", func() {
+		err = cs.CoreV1().Pods(ns).Delete(context.TODO(), serverPod1Name, metav1.DeleteOptions{})
+		framework.ExpectNoError(err, "failed to delete pod: %s on node", serverPod1Name)
+
+		err = <-errorChannel
```
reading the code it may happen that only the first pods answers all the UDP requests.
How can we guarantee in the test that maxAttemps lasts  enough to check both pods?
maybe signaling `continuousEcho` that we have a new pod and he has to send another round of udp packets?

JacobTanenbaum replied:
@aojea added a signal so continuousEcho() sends some packets waits and continues

JacobTanenbaum commented:
/retest

JacobTanenbaum commented:
/retest

aojea commented:
it's looking good to me , some suggestions here https://github.com/JacobTanenbaum/kubernetes/pull/1/files to add more comments, improve the logging and change the order of the UDP requests to start after the first pod is cleaned

JacobTanenbaum commented:
@aojea I made the logging changes. If you start the UDP traffic after the second pod starts the conntrack entries won't be created

aojea commented:
/lgtm
sorry for the loooong review :sweat_smile: , 
the output of the test now is great IMHO, in this run there was also 2 failures and we can observe them clearly, (maybe a hiccup during the pod creation/deletion)??
```
[BeforeEach] [sig-network] Services
  test/e2e/network/service.go:807
[It] should be able to preserve UDP traffic when server pod cycles for a NodePort service
  test/e2e/network/service.go:1242
[1mSTEP[0m: creating a UDP service clusterip-test with type=NodePort in services-8222
[1mSTEP[0m: creating a backend pod for the service clusterip-test
[1mSTEP[0m: checking NodePort service clusterip-test on node with public IP 172.18.0.2
May 12 16:00:10.547: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:12.585: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:14.553: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:16.556: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:18.583: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:20.577: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:22.553: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:24.553: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:26.565: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:28.555: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:30.556: INFO: The status of Pod server-1 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:32.587: INFO: The status of Pod server-1 is Running (Ready = true)
[1mSTEP[0m: Sending UDP traffic to NodePort service clusterip-test on node with publicIP 172.18.0.2
[1mSTEP[0m: creating a second pod for the service clusterip-test
May 12 16:00:32.659: INFO: The status of Pod server-2 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:33.587: INFO: Continuous echo ("udp://172.18.0.2:31429"): read udp 172.18.0.1:39911->172.18.0.2:31429: read: connection refused - 1 errors seen so far
May 12 16:00:34.717: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:34.821: INFO: The status of Pod server-2 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:35.724: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:36.678: INFO: The status of Pod server-2 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:36.737: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:37.737: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:37.737: INFO: Continuous echo waiting for signal to continue
May 12 16:00:38.688: INFO: The status of Pod server-2 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:40.667: INFO: The status of Pod server-2 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:42.694: INFO: The status of Pod server-2 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:44.674: INFO: The status of Pod server-2 is Pending, waiting for it to be Running (with Ready = true)
May 12 16:00:46.669: INFO: The status of Pod server-2 is Running (Ready = true)
May 12 16:00:46.669: INFO: Cleaning up server-1 pod
[1mSTEP[0m: Sending UDP traffic to NodePort service clusterip-test on node with publicIP 172.18.0.2
May 12 16:00:47.713: INFO: Continuous echo ("udp://172.18.0.2:31429"): read udp 172.18.0.1:39911->172.18.0.2:31429: read: connection refused - 2 errors seen so far
May 12 16:00:48.717: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:49.723: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:50.725: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:51.725: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:52.729: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:53.729: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:54.730: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:55.731: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:56.731: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:57.733: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:58.737: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:00:59.737: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:01:00.738: INFO: Continuous echo("udp://172.18.0.2:31429"): success
May 12 16:01:01.739: INFO: Continuous echo("udp://172.18.0.2:31429"): success
[AfterEach] [sig-network] Services
  test/e2e/framework/framework.go:175
May 12 16:01:01.739: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
[1mSTEP[0m: Destroying namespace "services-8222" for this suite.
[AfterEach] [sig-network] Services
  test/e2e/network/service.go:811


[32m• [SLOW TEST:52.413 seconds][0m
[sig-network] Services
[90mtest/e2e/network/framework.go:23[0m
  should be able to preserve UDP traffic when server pod cycles for a NodePort service
```
/assign @dcbw @danwinship

JacobTanenbaum commented:
/retest

JacobTanenbaum commented:
/retest

aojea commented:
/lgtm cancel
the test was flaky here 🤔 
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/90180/pull-kubernetes-e2e-gce/1260254585310679041/

JacobTanenbaum commented:
/retest

JacobTanenbaum commented:
/retest

aojea commented:
sorry I couldn't follow up, seems that the CI is flakier than usual but this test seems very sensitive to this, current failure is weird, it fails on the 10 first tries but succeed in the last ones

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/90180/pull-kubernetes-e2e-kind/1260295981765758977

```

[sig-network] Services should be able to preserve UDP traffic when server pod cycles for a NodePort service expand_less | 1m3s
-- | --
test/e2e/network/service.go:1242 May 12 20:26:58.776: pod communication failed Unexpected error:     <*errors.errorString \| 0xc002832260>: {         s: "continuous echo was not able to communicate with initial server pod",     }     continuous echo was not able to communicate with initial server pod occurred test/e2e/network/service.go:1297


[sig-network] Services should be able to preserve UDP traffic when server pod cycles for a NodePort service expand_less | 1m3s
-- | --
test/e2e/network/service.go:1242 May 12 20:26:58.776: pod communication failed Unexpected error:     <*errors.errorString \| 0xc002832260>: {         s: "continuous echo was not able to communicate with initial server pod",     }     continuous echo was not able to communicate with initial server pod occurred test/e2e/network/service.go:1297
```

JacobTanenbaum commented:
@aojea I don't see the success that you mention, continuousEcho() tries to connect to the initail server pod 10 times and fails with "connection refused" and when the signal comes that the second pod is ready because we never connected successfully to the first one we fail the test. Let me start up a cluster and see if it will reproduce locally maybe the last round of edits messed up the logic in some way

aojea commented:
> @aojea I don't see the success that you mention, continuousEcho() tries to connect to the initail server pod 10 times and fails with "connection refused" and when the signal comes that the second pod is ready because we never connected successfully to the first one we fail the test. Let me start up a cluster and see if it will reproduce locally maybe the last round of edits messed up the logic in some way

 you can't see it because it is not there, my fault, I was checking multiple logs at the same time, and it seems I mixed things

aojea commented:
> you can't see it because it is not there, my fault, I was checking multiple logs at the same time, and it seems I mixed things

oooh, it seems I didn't mix things, click in the failing test and it opens a new window with the logs of the failed test, then search for `INFO: Continuous echo("udp://172.18.0.4:31393"): success`

The info is being logged after the cleanup of the test

```
May 12 20:27:04.190: INFO: 	Container test-container ready: true, restart count 0
W0512 20:27:04.201344   11775 metrics_grabber.go:94] Master node is not registered. Grabbing metrics from Scheduler, ControllerManager and ClusterAutoscaler is disabled.
May 12 20:27:04.809: INFO: Continuous echo("udp://172.18.0.4:31393"): success
May 12 20:27:05.815: INFO: Continuous echo("udp://172.18.0.4:31393"): success
May 12 20:27:06.819: INFO: Continuous echo("udp://172.18.0.4:31393"): success
May 12 20:27:07.824: INFO: Continuous echo("udp://172.18.0.4:31393"): success
May 12 20:27:08.403: INFO: 
Latency metrics for node kind-worker
```

JacobTanenbaum commented:
@aojea good spot! it came from me making a mistake and not returning after not being able to connect to the initial backend pod, which is very curious why we could connect to the second backend but not the first

JacobTanenbaum commented:
/retest

JacobTanenbaum commented:
/retest

aojea commented on test/e2e/network/service.go:
```diff
@@ -518,6 +518,123 @@ func pokeUDP(host string, port int, request string, params *UDPPokeParams) UDPPo
+		}
+
+		bufsize := len(strconv.Itoa(i)) + 1
+		var buf = make([]byte, bufsize)
```
just curious, why do we have to recreate the buffer and increase the size in each iteration?

JacobTanenbaum replied:
oversight from when I created this function. Moved it to outside the for loop

JacobTanenbaum commented:
/retest

aojea commented:
/lgtm
let's see if it goes through this time

JacobTanenbaum commented:
/retest

dcbw commented:
/approve

dcbw commented:
/lgtm

JacobTanenbaum commented:
/retest
